application_marshalling_test.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2017 Rohith All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package marathon
+
+import (
+	"encoding/json"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestEnvironmentVariableUnmarshal(t *testing.T) {
+	defaultConfig := NewDefaultConfig()
+	configs := &configContainer{
+		client: &defaultConfig,
+		server: &serverConfig{
+			scope: "environment-variables",
+		},
+	}
+
+	endpoint := newFakeMarathonEndpoint(t, configs)
+	defer endpoint.Close()
+
+	application, err := endpoint.Client.Application(fakeAppName)
+	require.NoError(t, err)
+
+	env := application.Env
+	secrets := application.Secrets
+
+	require.NotNil(t, env)
+	assert.Equal(t, "bar", (*env)["FOO"])
+	assert.Equal(t, "TOP", (*secrets)["secret"].EnvVar)
+	assert.Equal(t, "/path/to/secret", (*secrets)["secret"].Source)
+}
+
+func TestMalformedPayloadUnmarshal(t *testing.T) {
+	var tests = []struct {
+		expected string
+		given    []byte
+	}{
+		{
+			expected: "unexpected secret field",
+			given:    []byte(`{"env": {"FOO": "bar", "SECRET": {"not_secret": "secret1"}}, "secrets": {"secret1": {"source": "/path/to/secret"}}}`),
+		},
+		{
+			expected: "unexpected secret field",
+			given:    []byte(`{"env": {"FOO": "bar", "SECRET": {"secret": 1}}, "secrets": {"secret1": {"source": "/path/to/secret"}}}`),
+		},
+		{
+			expected: "unexpected environment ",
+			given:    []byte(`{"env": {"FOO": 1, "SECRET": {"secret": "secret1"}}, "secrets": {"secret1": {"source": "/path/to/secret"}}}`),
+		},
+		{
+			expected: "malformed application d",
+			given:    []byte(`{"env": "value"}`),
+		},
+	}
+
+	for _, test := range tests {
+		tmpApp := new(Application)
+
+		err := json.Unmarshal(test.given, &tmpApp)
+		if assert.Error(t, err) {
+			assert.Equal(t, test.expected, err.Error()[0:23])
```

Comment:
Can we use `strings.HasPrefix` in the assertion to be less coupled to the actual character count? Feel then free to extend the `expected` strings as you deem helpful.

This will also save us from having to include the `expected` string into the message on line 79 (right now, if that line fails, we won't be able to tell which test case was hit).